dali/kernels/imgproc/convolution/convolution_cpu.h

```diff
@@ -0,0 +1,375 @@
+ public:
+  CyclicWindowWrapper(T* __restrict__ ptr, int length, int num_lanes = 1)
+      : data_(ptr), start_(0), end_(0), elements_(0), length_(length), num_lanes_(num_lanes) {
+    DALI_ENFORCE(num_lanes <= max_lanes, "Static lanes limit exceeded.");
+  }
+
+  /**
+   * @brief Drop one element consistine of `NumLanes()` lanes from the buffer, moving the start
+   *        element.
+   */
+  void PopElement() {
+    assert(elements_ > 0);
+    elements_--;
+    start_++;
+    WrapPosition(start_);
+  }
+
+  /**
+   * @brief Add a window element, that has `NumLanes()` consecutive lanes and starting at `input`,
+   * to the buffer.
+   */
+  void PushElement(const T* __restrict__ input) {
+    assert(elements_ < length_);
+    memcpy(data_ + end_ * NumLanes(), input, sizeof(T) * NumLanes());
+    elements_++;
+    end_++;
+    WrapPosition(end_);
+  }
+
+  /**
+   * @brief Add a window element to the buffer, that has `NumLanes()` lanes
+   */
+  void PushElement(span<const T> input) {
+    assert(elements_ < length_);
+    for (int c = 0; c < NumLanes(); c++) {
+      data_[end_ * NumLanes() + c] = input[c];
+    }
+    elements_++;
+    end_++;
+    WrapPosition(end_);
+  }
+
+  /**
+   * @brief Get pointer to the start of pixel `idx` in the buffer, taking into account the cyclic
+   * wrapping
+   */
+  T* GetElementOffset(int idx) {
+    assert(idx < elements_);
+    if (start_ + idx < length_) {
+      return data_ + (start_ + idx) * NumLanes();
+    } else {
+      return data_ + (start_ + idx - length_) * NumLanes();
+    }
+  }
+
+  /**
+   * @brief Calculate dot product with 1-channel `window` length equal to the one specified
+   * at construction. The result is stored in pixel stored at `accum`.
+   */
+  template <typename W>
+  void CalculateDot(W* __restrict__ accum, const W* __restrict__ window) {
+    assert(elements_ == length_);
+    for (int c = 0; c < NumLanes(); c++) {
+      accum[c] = 0;
+    }
+    int window_idx = 0;
+    for (int buf_idx = start_; buf_idx < length_; buf_idx++, window_idx++) {
+      for (int c = 0; c < NumLanes(); c++) {
+        accum[c] += window[window_idx] * data_[buf_idx * NumLanes() + c];
+      }
+    }
+    for (int buf_idx = 0; buf_idx < end_; buf_idx++, window_idx++) {
+      for (int c = 0; c < NumLanes(); c++) {
+        accum[c] += window[window_idx] * data_[buf_idx * NumLanes() + c];
+      }
+    }
+  }
+
+  template <typename U, typename W>
+  void CalculateDot(U* __restrict__ output, const W* __restrict__ window, W scale) {
+    std::array<W, max_lanes> tmp;
+    CalculateDot(tmp.data(), window);
+    for (int c = 0; c < NumLanes(); c++) {
+      output[c] = ConvertSat<U>(tmp[c] * scale);
+    }
+  }
+
+  int Size() {
+    return elements_;
+  }
+
+  bool Empty() {
+    return elements_ == 0;
+  }
+
+ private:
+  void WrapPosition(int& pos) {
+    if (pos == length_) {
+      pos = 0;
+    }
+  }
+
+  int NumLanes() {
+    return std::min(num_lanes_, max_lanes);
```

Comment:
It's already enforcing it in the constructor... is this function even necessary?
```suggestion
    return num_lanes_;
```

Reply:
My intention was to hint the compiler, that whenever I use the num_lanes_ it's less than max_lanes. But I didn't test if that helps much.

Reply:
Maybe if it got unrolled... but I don't think any compiler would be that smart.